rspotify-model/src/idtypes.rs

```diff
@@ -380,111 +420,101 @@ macro_rules! define_impls {
                     write!(f, "{}", self.uri())
                 }
             }
-
-            /// IDs can also be used to convert from a `str`; this works both
-            /// with IDs and URIs.
-            impl std::str::FromStr for $name {
-                type Err = IdError;
-
-                fn from_str(s: &str) -> Result<Self, Self::Err> {
-                    Self::from_id_or_uri(s)
-                }
-            }
         )+
     }
 }
 
 // First declaring the regular IDs. Those with custom behaviour will have to be
 // declared manually later on.
 define_idtypes!(
-    Artist => ArtistId,
-    Album => AlbumId,
-    Track => TrackId,
-    Playlist => PlaylistId,
-    Show => ShowId,
-    Episode => EpisodeId
-);
-
-/// ID of type [`Type::User`]. Refer to the [module-level
-/// docs][`crate::idtypes`] for more information.
-///
-/// Note that the implementation of this specific ID differs from the rest: a
-/// user's ID doesn't necessarily have to be made of alphanumeric characters. It
-/// can also use underscores and other characters, but since Spotify doesn't
-/// specify it explicitly, this just allows any string as an ID.
-#[derive(Clone, Debug, PartialEq, Eq, Serialize, Hash)]
-pub struct UserId(String);
-impl Id for UserId {
-    #[inline]
-    fn id(&self) -> &str {
-        &self.0
-    }
-
-    #[inline]
-    fn _type(&self) -> Type {
-        Type::User
-    }
-
-    #[inline]
-    fn _type_static() -> Type
-    where
-        Self: Sized,
-    {
-        Type::User
+    Artist => {
+        name: ArtistId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    Album => {
+        name: AlbumId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    Track => {
+        name: TrackId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    Playlist => {
+        name: PlaylistId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    Show => {
+        name: ShowId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    Episode => {
+        name: EpisodeId,
+        validity: |id| id.chars().all(|ch| ch.is_ascii_alphanumeric())
+    },
+    User => {
+        name: UserId,
+        validity: |_| true
     }
+);
 
-    #[inline]
-    unsafe fn from_id_unchecked(id: &str) -> Self {
-        UserId(id.to_owned())
+// We use `enum_dispatch` for dynamic dispatch, which is not only easier to use
+// than `dyn`, but also more efficient.
+/// Grouping up multiple kinds of IDs to treat them generically. This also
+/// implements [`Id`], and [`From`] to instantiate it.
+#[enum_dispatch(Id)]
+pub enum PlayContextId<'a> {
+    Artist(ArtistId<'a>),
+    Album(AlbumId<'a>),
+    Playlist(PlaylistId<'a>),
+    Show(ShowId<'a>),
+}
+// These don't work with `enum_dispatch`, unfortunately.
+impl<'a> PlayContextId<'a> {
+    pub fn as_ref(&'a self) -> PlayContextId<'a> {
```

Comment:
I am a little confused about how this `as_ref` function works?

Reply:
It makes it possible to reborrow the internal string so that no cloning is necessary. It just takes another immutable reference to the inner contents, with the same lifetime.

This is what I'm the most unhappy about in the PR. It is somewhat annoying to use but I think it's the only way.